libsolidity/interface/FileReader.h

```diff
@@ -0,0 +1,87 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <boost/filesystem.hpp>
+#include <map>
+#include <vector>
+
+#include <libsolidity/interface/ReadFile.h>
+
+namespace solidity {
+
+class FileReader
+{
+public:
+	using StringMap = std::map<std::string, std::string>;
+
+	FileReader(
+		boost::filesystem::path _basePath,
+		std::vector<boost::filesystem::path> _allowedDirectories
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_fullPathMapping()
+	{}
+
+	FileReader(): FileReader({}, {}) {}
+
+	boost::filesystem::path const& basePath() const noexcept { return m_basePath; }
+	void setBasePath(boost::filesystem::path const& _path);
+
+	void allowDirectory(boost::filesystem::path _path);
+
+	std::vector<boost::filesystem::path> const& allowedDirectories() const noexcept { return m_allowedDirectories; }
+
+	StringMap const& sourceCodes() const noexcept { return m_sourceCodes; }
+	StringMap& sourceCodes() noexcept { return m_sourceCodes; }
```

Comment:
Why allow messing from the sources from the outside? At first I thought it might be because of `CompilerStack::setSources()` but that method expects `StringMap`, not a mutable reference.

Reply:
I am pretty sure I used that to move the map out of this, as FileReader() wasn't used after that anymore. But the code has changed, so I'll remove this.

Reply:
Oh, It's CommandLineInterface.cpp's link() function that does actually still mutate `m_sourceCodes` from the outside.

It would be worked around by just retrieving the copy, modifying the copy, and copying then back. But that sounds expensive. Any ideas?

Reply:
Maybe you could move it out, link it and then move it back in?

```c++
StringMap relinquishSourceCodes() noexcept {
    StringMap tmpMap = move(m_sourceCodes);
    m_sourceCodes = {};
    return tmpMap;
}
```

The downside is that you would have to either keep other members in sync or remember that the reader is in an unusable state in the meantime.

Another option would be to just built the source code map first, link it and only then give it to the file reader.